plugin/plugins/filesystem/nodes/base.go

```diff
@@ -0,0 +1,275 @@
+package fsnodes
+
+import (
+	"context"
+	"errors"
+	gopath "path"
+	"time"
+
+	"github.com/djdv/p9/p9"
+	"github.com/djdv/p9/unimplfs"
+	nodeopts "github.com/ipfs/go-ipfs/plugin/plugins/filesystem/nodes/options"
+	logging "github.com/ipfs/go-log"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	corepath "github.com/ipfs/interface-go-ipfs-core/path"
+)
+
+const ( //device - attempts to comply with standard multicodec table
+	dMemory = 0x2f // generic "path"
+	dIPFS   = 0xe4
+)
+
+var _ p9.File = (*Base)(nil)
+
+type p9Path = uint64
+
+// Base provides a foundation to build file system nodes which contain file meta data
+// as well as stubs for unimplemented file system methods
+type Base struct {
+	// Provide stubs for unimplemented methods
+	unimplfs.NoopFile
+	p9.DefaultWalkGetAttr
+
+	Trail []string // FS "breadcrumb" trail from node's root
+
+	// Storage for file's metadata
+	Qid      *p9.QID
+	meta     *p9.Attr
+	metaMask *p9.AttrMask
+	Logger   logging.EventLogger
+
+	closed   bool // set to true upon close; this reference should not be used again for anything
+	modified bool // set to true when the `Trail` has been modified (usually by `Step`)
+	// reset to false when `Qid` has been populated with the current path in `Trail` (usually by `QID`)
+
+}
+
+func newBase(ops ...nodeopts.AttachOption) Base {
+	options := nodeopts.AttachOps(ops...)
+
+	return Base{
+		Logger:   options.Logger,
+		Qid:      new(p9.QID),
+		meta:     new(p9.Attr),
+		metaMask: new(p9.AttrMask),
+	}
+}
+
+func (b *Base) clone() Base {
+	return *b
+}
+
+func (b *Base) Fork() Base {
+	newFid := b.clone()
+	newFid.Trail = make([]string, len(b.Trail)+1) // NOTE: +1 is preallocated space for `Step`; not required
```

Comment:
Yeah, that was wrong. Instead of anticipating a step, we now just copy using the 2nd method, and let append handle allocation during `Step`.

https://github.com/ipfs/go-ipfs/pull/6612/commits/9ef8e6ad9c4fc3f0213cea2836e9baf585c7987e#diff-6c2c294fc60db12a68ce40457ff3ea5eL64
https://github.com/ipfs/go-ipfs/pull/6612/commits/9ef8e6ad9c4fc3f0213cea2836e9baf585c7987e#diff-6c2c294fc60db12a68ce40457ff3ea5eL249

The intent here is that `Fork` should create a `newFid` that contains the same path as `oldFid`, but is also ready to `Walk`.
***
Related: I extracted the walk logic and changed the docs
https://github.com/ipfs/go-ipfs/pull/6612/commits/9ef8e6ad9c4fc3f0213cea2836e9baf585c7987e#diff-12cb59aacb05b8b81e16c020457e05a6

Additional context: (maybe this should go into the source comments somewhere?)
In an attempt to reduce forking for each path component, we allow `Step` to change the fids path directly.
For our implementations we do this when paths are withing the same filesystem.
Deriving a new node only when we escape from one filesystem to another.

e.g. `ipfsDir := root.Walk([]"ipfs", "Qm...")` derives a new reference to `root` that is ready to `Step`.
`root` steps to "ipfs" by returning a derivative of the `PinFS` it's attached to.
`PinFS` then steps to "Qm..." by returning a derivative of the `IPFS` it's attached to, and this reference is returned to the caller.

`clone := ipfsDir.Walk(nil)` would derive a new reference of the `ipfsDir`, ready to be `Walk`.

and at this point `clone.Walk("subDir", "deeper")` only needs to modify the fids `Trail` without needing to derive new references anymore.